### ISSUE TITLE
Help in the file DTMFUtil.java 

### ISSUE THREAD

**Issue by VibhavChaddha** [tino1b2be/DTMF-Decoder#2]
There is a lot of code to grasp quickly and then to understand the flow of the code. I looked into the 'DTMFDecoder.java' file, but when I was running it, there was no output. It said "The DTMF tones found in the given file are: ", that's it. Although I have not looked deeper into it, but I think the 'sequence' variable is not getting any value.
Thanks.

**Comment by tino1b2be** [comment]
Hey there. `DTMFUtil.java` has no "main method" (by main method I assume you are referring to the actual main method that is used to run a java program). I just updated the readme with examples of how to use the decoder. You can check out [DTMFDecoder.java](https://github.com/tino1b2be/DTMF-Decoder/blob/master/source/com/tino1b2be/cmdprograms/DTMFDecoder.java) which is a small program that actually uses the decoder. There are other programs and tests inside [/cmdprograms](https://github.com/tino1b2be/DTMF-Decoder/tree/master/source/com/tino1b2be/cmdprograms).

**Comment by VibhavChaddha** [comment]
Yeah, I looked into the 'DTMFDecoder.java' file, but when I was running it, there was no output. It said "The DTMF tones found in the given file are: ", that's it. Although I have not looked deeper into it, but I think the 'sequence' variable is not getting any value. So I thought of asking you for help.
I am actually quite new to Java audio, so I'm asking all these out of pure ignorance.

**Comment by tino1b2be** [comment]
Ohh. maybe the audio file you are decoding doesnt have any DTMF tones in it? Try testing with one of the audio files I have in [this folder](https://github.com/tino1b2be/DTMF-Decoder/tree/master/samples)

Disclaimer: When I wrote this program I had very limited knowledge about signal detection techniques (2nd year electrical engineering) so maybe that's why it's failing to detect but I did make sure it could detect DTMF tones as specified by the [ITU-T Q.24](https://www.itu.int/rec/dologin_pub.asp?lang=e&id=T-REC-Q.24-198811-I!!PDF-E&type=items)

**Comment by VibhavChaddha** [comment]
Your code is brilliantly choreographed and quite easy to understand. But just because of the length of the code I tend to forget few things, like 'where was I before this part'.
I will have to make a hell lot of changes in your code to fulfil my purpose. I have got a few understanding of this code, though. Good for me.

What I exactly wanted is that "I have a wave file with two frequencies, and whenever I find the first frequency in the file I will write that in a binary array as '0' and I will write the second frequency in the same file as '1'. For example, if my wave file is 10 seconds long and for the first 5 seconds it contains the first frequency and next 5 seconds it contains second frequency, then my array should be '0000011111' ".
I want to achieve it using FFT, as what I have read in past few weeks, Goertzel will not work perfectly if there is some noise.

**Comment by tino1b2be** [comment]
Thank you 👍 💃 :D ... Yeah I'm sorry about the code bloat there. If you study the code more I'm sure you can modify it for your purpose. 

I also had the same problems with the Goertzel algorithm. It wasn't very good when there was a lot of noise. The FFT from apache commons math library is actually WAY better than the first order goertzel algorithm i implemented because it's optimised and therefore much faster (Refer to [my report on page 14](https://github.com/tino1b2be/DTMF-Decoder/blob/master/Documentation/DTMF%20Decoder%20Report.pdf)). 

I'm not sure how much experience you have with DSP but you can try implement a window function and/or a matched filter (or some other sort of correlation if you know the exact frequency(s) you are looking for) to improve detection performance when there's a lot of noise.

Check out this [small program](https://github.com/tino1b2be/DTMF-Decoder/blob/master/source/com/tino1b2be/cmdprograms/TryFFTSpectrum.java) it can give you a head start for working with FFT.

**Comment by tino1b2be** [comment]
If I may ask, how did you come across this repo and what what are you using it for? (is it a varisty project? personal use? etc...)

**Comment by VibhavChaddha** [comment]
Hi.
There was some issue with my internet yesterday and therefore I accidentally deleted my comments.
I was searching for some libraries which I can use for my project and I found your repo. 
I am actually using it for my personal use right now, but maybe in future, it will help me in some way or the other, as knowledge never goes waste. :)
Thanks a lot for taking your valuable time to help me out with my issues. I really appreciate it and I am definitely going to look at your FFT code for reference.
Hopefully, I will be in touch with you. Cheerio.

**Comment by VibhavChaddha** [comment]
My yesterdays' comment. ;)

Your code is brilliantly choreographed and quite easy to understand. But just because of the length of the code I tend to forget few things, like 'where was I before this part'.
I will have to make a hell lot of changes in your code to fulfil my purpose. I have got a few understanding of this code, though. Good for me.

What I exactly wanted is that "I have a wave file with two frequencies, and whenever I find the first frequency in the file I will write that in a binary array as '0' and I will write the second frequency in the same file as '1'. For example, if my wave file is 10 seconds long and for the first 5 seconds it contains the first frequency and next 5 seconds it contains second frequency, then my array should be '0000011111' ".
I want to achieve it using FFT, as what I have read in past few weeks, Goertzel will not work perfectly if there is some noise.

**Comment by VibhavChaddha** [comment]
Hi, tino1b2be.

In your code [TryFFTSpectrum](uhttps://github.com/tino1b2be/DTMF-Decoder/blob/master/source/com/tino1b2be/cmdprograms/TryFFTSpectrum.javarl) the power that you are printing, what exactly is that? Is that the peak in the amplitude when the specified frequency is found? And is there any guarantee of it being always constant (or just a slight change) for a particular frequency?

**Comment by tino1b2be** [comment]
Let me check it out I last used it over a yer ago... I'll modify if necessary.

**Comment by VibhavChaddha** [comment]
Sure. Thanks.

**Comment by VibhavChaddha** [comment]
Hi, tino1b2be. Wish you a very Happy New Year.
I saw that you have made some amendments into 'TryFFTSpectrum.java' file. Still, I have the above-mentioned query that, "the power that you are printing, what exactly is that? Is that the peak in the amplitude when the specified frequency is found?". Can you please help me out with it?

P.S. - I want to check for frequencies between 16000 to 20000.
Thanks.

**Comment by tino1b2be** [comment]
Closing this one, it's been dormant since 2017 and was a general DSP question rather than a bug in the decoder. For anyone landing here later: the `power` printed by `TryFFTSpectrum` is the magnitude of the FFT bin for the queried frequency (see the report in `/Documentation` for the derivation). Feel free to open a new issue if you hit an actual bug.
